File: packages/go/contracts/domain/strategy.go
package domain

import "time"

type StrategyID string

type StrategyState string

type SignalAction string

const (
	StrategyStateDraft    StrategyState = "draft"
	StrategyStateActive   StrategyState = "active"
	StrategyStateArchived StrategyState = "archived"

	SignalActionEnterLong  SignalAction = "enter_long"
	SignalActionExitLong   SignalAction = "exit_long"
	SignalActionEnterShort SignalAction = "enter_short"
	SignalActionExitShort  SignalAction = "exit_short"
	SignalActionHold       SignalAction = "hold"
)

type StrategyDefinition struct {
	ID            StrategyID     `json:"id"`
	Name          string         `json:"name"`
	Version       string         `json:"version"`
	State         StrategyState  `json:"state"`
	InstrumentIDs []InstrumentID `json:"instrument_ids,omitempty"`
	IndicatorRefs []string       `json:"indicator_refs,omitempty"`
	Parameters    map[string]any `json:"parameters,omitempty"`
}

type StrategySignal struct {
	StrategyID   StrategyID     `json:"strategy_id"`
	InstrumentID InstrumentID   `json:"instrument_id"`
	Action       SignalAction   `json:"action"`
	Timestamp    time.Time      `json:"timestamp"`
	Strength     Decimal        `json:"strength,omitempty"`
	Metadata     map[string]any `json:"metadata,omitempty"`
}

File: packages/python/src/trend_buddy_shared/contracts/domain/strategy.py
from dataclasses import dataclass, field
from datetime import datetime
from decimal import Decimal
from enum import StrEnum
from typing import Any


class StrategyState(StrEnum):
    DRAFT = "draft"
    ACTIVE = "active"
    ARCHIVED = "archived"


class SignalAction(StrEnum):
    ENTER_LONG = "enter_long"
    EXIT_LONG = "exit_long"
    ENTER_SHORT = "enter_short"
    EXIT_SHORT = "exit_short"
    HOLD = "hold"


@dataclass
class StrategyDefinition:
    id: str
    name: str
    version: str
    state: StrategyState
    instrument_ids: list[str] = field(default_factory=list)
    indicator_refs: list[str] = field(default_factory=list)
    parameters: dict[str, Any] = field(default_factory=dict)


@dataclass
class StrategySignal:
    strategy_id: str
    instrument_id: str
    action: SignalAction
    timestamp: datetime
    strength: Decimal = Decimal("0")
    metadata: dict[str, Any] = field(default_factory=dict)
